### PR TITLE
[MFT] Standalone track fitting

### PR DESCRIPTION
**Pull request by rpezzi**
This is the first version of MFT standalone track fitting. Fitting methods were adapted from MCH. The coordinate system has been adapted to the MFT: (X, Y, Phi, Tanl, q/pt). Further improvements are planned: refine track seed covariance matrix, add MS effects, create an alternative to TrackParCov class suitable for MFT tracks in this coordinate system; create a MUON base cluster class.

### CODE REVIEW

**Comment by bovulpes** [comment]
It is ok for me to merge this PR.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ private:
+  Double_t mZ = 0.; ///< Z coordinate (cm)
+
+  /// Track parameters ordered as follow:      <pre>
+  /// X       = X coordinate   (cm)
+  /// Y       = Y coordinate   (cm)
+  /// PHI     = azimutal angle
+  /// TANL    = tangent of \lambda (dip angle)
+  /// INVQPT    = Inverse transverse momentum (GeV/c ** -1) times charge (assumed forward motion)  </pre>
+  TMatrixD mParameters{5, 1}; ///< \brief Track parameters
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h]: Why using TMatrixD, even for internal use? It derives from the TObject so has all the virtuality overheads. Also, TMatrixD is an arbitrary dim. matrix, while your dimensions are fixed. Why don't you use a std::array here (or, if you want to use matrix operations, SVector class)?
Also, do you really need double?

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+  /// Covariance matrix of track parameters, ordered as follows:    <pre>
+  ///  <X,X>         <Y,X>           <PHI,X>       <TANL,X>        <INVQPT,X>
+  ///  <X,Y>         <Y,Y>           <PHI,Y>       <TANL,Y>        <INVQPT,Y>
+  /// <X,PHI>       <Y,PHI>         <PHI,PHI>     <TANL,PHI>      <INVQPT,PHI>
+  /// <X,TANL>      <Y,TANL>       <PHI,TANL>     <TANL,TANL>     <INVQPT,TANL>
+  /// <X,INVQPT>   <Y,INVQPT>     <PHI,INVQPT>   <TANL,INVQPT>   <INVQPT,INVQPT>  </pre>
+  mutable std::unique_ptr<TMatrixD> mCovariances{}; ///< \brief Covariance matrix of track parameters
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h]: The cov.matrix is symmetric, either use array of 15 values (preferable) or ``ROOT::Math::SMatrix<double, 5, 5, ROOT::Math::MatRepSym<double, 5>>;``, avoid TMatrixD at any rate!

File: Detectors/ITSMFT/MFT/tracking/src/TrackExtrap.cxx [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+  double phi0 = trackParam->getPhi();
+  double tanl0 = trackParam->getTanl();
+  double invqpt0 = trackParam->getInvQPt();
+  double dZ2 = dZ * dZ;
+  double cosphi0 = TMath::Cos(phi0);
+  double sinphi0 = TMath::Sin(phi0);
+  double tanl0sq = tanl0 * tanl0;
+  double k = -mBZField * 0.3;
+
+  TMatrixD jacob(5, 5);
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/tracking/src/TrackExtrap.cxx]: idem, you can use here a fixed size matrix, no need for TMatrixD

**Comment by ktf** [comment]
Please fix the code formatting.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackExtrap.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  TrackExtrap() = default;
+  ~TrackExtrap() = default;
+
+  TrackExtrap(const TrackExtrap&) = delete;
+  TrackExtrap& operator=(const TrackExtrap&) = delete;
+  TrackExtrap(TrackExtrap&&) = delete;
+  TrackExtrap& operator=(TrackExtrap&&) = delete;
+
+  void setBz(float bZ) { mBZField = bZ; } /// Set the magnetic field for the MFT
+  float getBz() { return mBZField; }
```
[the remark this answers is not in the document]
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackExtrap.h]: Done

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackExtrap.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  static bool extrapToZ(TrackParam* trackParam, double zEnd, bool isFieldON = true);
+  static bool extrapToZCov(TrackParam* trackParam, double zEnd, bool updatePropagator = false, bool isFieldON = true);
+  static void linearExtrapToZ(TrackParam* trackParam, double zEnd);
+  static void linearExtrapToZCov(TrackParam* trackParam, double zEnd, bool updatePropagator);
+  static void helixExtrapToZ(TrackParam* trackParam, double zEnd);
+  static void helixExtrapToZCov(TrackParam* trackParam, double zEnd, bool updatePropagator);
+  static void addMCSEffect(TrackParam* trackParam, double dZ, double x0, bool isFieldON = true);
+
+ private:
+  static Float_t mBZField; // Tesla.
```
[the remark this answers is not in the document]
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackExtrap.h]: Fixed

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackFitter.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  TrackFitter& operator=(const TrackFitter&) = delete;
+  TrackFitter(TrackFitter&&) = delete;
+  TrackFitter& operator=(TrackFitter&&) = delete;
+
+  void setBz(float bZ);
+
+  /// Enable/disable the smoother (and the saving of related parameters)
+  void smoothTracks(bool smooth) { mSmooth = smooth; }
+  /// Return the smoother enable/disable flag
+  bool isSmootherEnabled() { return mSmooth; }
```
[the remark this answers is not in the document]
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackFitter.h]: Fixed

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackFitter.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+ private:
+  void initTrack(const o2::itsmft::Cluster& cl, const o2::itsmft::Cluster& cl2, TrackParam& param);
+  void addCluster(const TrackParam& startingParam, const o2::itsmft::Cluster& cl, TrackParam& param);
+  void smoothTrack(FitterTrackMFT& track, bool finalize);
+  void runSmoother(const TrackParam& previousParam, TrackParam& param);
+  Float_t mBZField = 0.5;                   // Tesla.
+  static constexpr double SMaxChi2 = 2.e10; ///< maximum chi2 above which the track can be considered as abnormal
+  /// z position of the layers
+  static constexpr float SDefaultLayerZ[10] = {-45.3, -46.7, -48.6, -50.0, -52.4, -53.8, -67.7, -69.1, -76.1, -77.5};
```
[the remark this answers is not in the document]
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackFitter.h]: Fixed. Now using o2::mft::constants from `MFTBase/Constants.h`

**Comment by rpezzi** [comment]
Dear @sawenzel  and @shahor02 , thanks for your inputs. I have addressed most of your comments. Particularly relevant is the creation of a new track class to store the MFT tracks. This class is derived from the track fitting parameter class, which will be trimmed once we decide if this coordinate system is successful. For proper assessment of the fitting quality the MC labels need to be propagated (TODO). As for the use of TMatrixD, we are using the same methods as the MCH already in O2@dev. I rather postpone this optimization until we decide on a common approach.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h (hunk 1 above)
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h]: See comment below.

File: Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h (hunk 2 above)
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/include/MFTTracking/TrackParam.h]: See comment bellow. SMatrix will most certainly be used.

File: Detectors/ITSMFT/MFT/tracking/src/TrackExtrap.cxx (hunk 3 above)
Comment (rpezzi) [reply Detectors/ITSMFT/MFT/tracking/src/TrackExtrap.cxx]: idem

File: DataFormats/Detectors/ITSMFT/MFT/include/DataFormatsMFT/TrackParamMFT.h [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+  const o2::itsmft::Cluster* mClusterPtr = nullptr; ///< Pointer to the associated cluster if any
+
+  Bool_t mRemovable = false; ///< kTRUE if the associated cluster can be removed from the track it belongs to
+
+  Double_t mTrackChi2 = 0.; ///< Chi2 of the track when the associated cluster was attached
+  Double_t mLocalChi2 = 0.; ///< Local chi2 of the associated cluster with respect to the track
+};
+
+class TrackMFTParam : public TrackParamMFT
```
Comment (shahor02) [inline DataFormats/Detectors/ITSMFT/MFT/include/DataFormatsMFT/TrackParamMFT.h]: deriving TrackMFTParam from TrackParamMFT is somewhat confusing, perhaps more appropriate names can be used? But I don't see  where the TrackMFTParam is used at all. Is this class needed?

File: DataFormats/Detectors/ITSMFT/MFT/src/TrackParamMFT.cxx [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+//#include "MFTTracking/Cluster.h"
+
+namespace o2
+{
+namespace mft
+{
+
+using namespace std;
+
+//_________________________________________________________________________
```
Comment (shahor02) [inline DataFormats/Detectors/ITSMFT/MFT/src/TrackParamMFT.cxx]: As far as I understand, the TrackParamMFT is supposed to be used only for internal, while the TrackMFT will get its own data members after the validation of new track model, right? 
Then, I think the TrackParamMFT should not be in the DataFormats directory since there we collect classes which are supposed to be messaged over the DPL. I.e. the TrackMFT should stay in the DataFormats and become asap the meseagable (POD) class, while the TrackParamMFT it is better to move to the MFT algorithms directory. Also, since at the moment it is a base class of the TrackMFT which must be persistent, I think you should add ClassDefNV to it.

Comment (rpezzi) [reply DataFormats/Detectors/ITSMFT/MFT/src/TrackParamMFT.cxx]: Yes, Ruben, you are right. The inheritance from TrackParamMFT was a dirty solution to get the track at the dataformat directory. I will follow your suggestion.

Comment (shahor02) [reply DataFormats/Detectors/ITSMFT/MFT/src/TrackParamMFT.cxx]: If due to the dependencies you need to keep the TrackParamMFT in the DataFormats, we can keep it at the moment there until you detach them. Anyway, if you keep the inheritance, the TrackMFT will not be messageable.

Comment (rpezzi) [reply DataFormats/Detectors/ITSMFT/MFT/src/TrackParamMFT.cxx]: Hi @shahor02 , TrackMFT has been cleaned-up and TrackParamMFT is back to where it belongs on the MFT Tracking directory.

File: Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx [hunk 10]
```diff
@@ -129,21 +129,21 @@ void TrackerDPL::run(ProcessingContext& pc)
         LOG(INFO) << "ROframe: " << roFrame << ", clusters loaded : " << nclUsed;
         mTracker->setROFrame(roFrame);
         mTracker->clustersToTracks(event);
-        tracks.swap(mTracker->getTracks());
+        //tracks.swap(mTracker->getTracks());
```
Comment (ktf) [inline Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx]: Please do not comment code. Remove if not needed.

Comment (rpezzi) [reply Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx]: Hi @ktf , I reviewed the code and removed some commented code. I am keeping those used for debugging.

File: Detectors/ITSMFT/MFT/workflow/src/TrackFitterSpec.cxx [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+//_________________________________________________________________________________________________
+void TrackFitterTask::run(ProcessingContext& pc)
+{
+
+  if (mState != 1)
+    return;
+
+  auto tracksLTF = pc.inputs().get<const std::vector<o2::mft::TrackLTF>>("tracksltf");
+  auto tracksCA = pc.inputs().get<const std::vector<o2::mft::TrackCA>>("tracksca");
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/workflow/src/TrackFitterSpec.cxx]: both TrackCA and TrackLTF are POD classes, instead of ``get<const std::vector<...>>`` one should use ``get<gsl::span<...>>`` (one can do this in separate PR).

File: Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx [hunk 12]
```diff
@@ -92,10 +92,10 @@ void TrackerDPL::run(ProcessingContext& pc)
               << mc2rofs.size() << " MC events";
   }
 
-  std::vector<o2::mft::TrackMFTExt> tracks;
+  //std::vector<o2::mft::TrackMFTExt> tracks;
```
Comment (shahor02) [inline Detectors/ITSMFT/MFT/workflow/src/TrackerSpec.cxx]: Also in this device the clusters are extracted (L73,74) via vectors (i.e. copy is created) instead of accessing them via span. 
Since this is out of scope of this PR, I propose that we merge it, but then please open another PR changing vectors to spans.
